source/kernel/src/lib.rs

```diff
@@ -197,41 +139,32 @@ impl Kernel {
 
         // process mailbox messages
         let inner = self.inner();
-        let inner_mut = self.inner_mut().unwrap();
         let u2k_buf: *mut BBBuffer = &self.inner.u2k_ring as *const _ as *mut _;
         let k2u_buf: *mut BBBuffer = &self.inner.k2u_ring as *const _ as *mut _;
         let u2k: FrameConsumer<'static> = unsafe { BBBuffer::take_framed_consumer(u2k_buf) };
-        let k2u: FrameProducer<'static> = unsafe { BBBuffer::take_framed_producer(k2u_buf) };
-
-        // Incoming messages
-        while let Some(msg) = u2k.read() {
-            match postcard::from_bytes::<UserRequest>(&msg) {
-                Ok(req) => {
-                    let kind = req.driver_kind();
-                    if let Some(drv) = inner_mut.drivers.iter().find(|drv| drv.kind == kind) {
-                        drv.queue
-                            .enqueue_sync(Message {
-                                request: req,
-                                response: inner.user_reply.clone(),
-                            })
-                            .map_err(drop)
-                            .unwrap();
+        let _k2u: FrameProducer<'static> = unsafe { BBBuffer::take_framed_producer(k2u_buf) };
+
+        #[allow(unreachable_code)]
+        if let Some(mut _reg) = self.registry.try_lock() {
+            // Incoming messages
+            while let Some(msg) = u2k.read() {
+                match postcard::from_bytes::<UserRequest>(&msg) {
+                    Ok(_req) => {
+                        // let kind = req.driver_kind();
+                        // if let Some(drv) = inner_mut.drivers.iter().find(|drv| drv.kind == kind) {
```

Comment:
*for now*, the number of drivers is likely to be small, and the "search" mechanism is an implementation detail.

I'd be open to a tree, or keeping the list sorted on insertion and using a binary search or something (since it still lives in linear memory).